src/index.spec.ts

```diff
@@ -4,9 +4,195 @@ import {
+    expect(toLowerCase('')).toBe('Please provide a string');
+  });
+
+  test('it should return an error message with null', () => {
+    expect(toLowerCase(null)).toBe('Please provide a string');
+  });
+});
+
+describe('removeDuplicatesFromArray tests', () => {
+  test('it should return the same array', () => {
+    const array: string[] = ['a', 'b', 'c'];
+    expect(removeDuplicatesFromArray(array)).toEqual(['a', 'b', 'c']);
+  });
+
+  test('it should return the same array with one element', () => {
+    const array: string[] = ['1'];
+    expect(removeDuplicatesFromArray(array)).toEqual(['1']);
+  });
+
+  test('it should return the same array with an array of differents types', () => {
+    const array: (string | number)[] = ['a', 1, 'b', 2];
+    expect(removeDuplicatesFromArray(array)).toEqual(['a', 1, 'b', 2]);
+  });
+
+  test('it should return an array without duplicates', () => {
+    const array: string[] = ['a', 'a', 'b', 'c'];
+    expect(removeDuplicatesFromArray(array)).toEqual(['a', 'b', 'c']);
+  });
+
+  test('it should throw an error message with null', () => {
+    const array: string[] = null;
+    expect(() => {
+      removeDuplicatesFromArray(array);
+    }).toThrow(new Error('please provide an array of numbers or strings'));
+  });
+});
+
+describe('createProduct tests', () => {
+  const validProduct = {
+    name: 'laptop',
+    tags: ['technology'],
+    description: 'Asus 13 inch',
+    price: 450.0,
+  };
+
+  test('with a valid product it should return the same product with an id', () => {
+    expect(createProduct(validProduct)).toEqual({
+      id: expect.any(Number),
+      ...validProduct,
+    });
+  });
+
+  test('it should throw an error with a name attribute of size 2', () => {
+    const invalidProduct = { ...validProduct };
+    invalidProduct['name'] = 'it';
+    expect(() => {
+      createProduct(invalidProduct);
+    }).toThrowErrorMatchingSnapshot();
+  });
+
+  test('it should throw an error with a description attribute of size 2', () => {
+    const invalidProduct = { ...validProduct };
+    invalidProduct['description'] = 'me';
+    expect(() => {
+      createProduct(invalidProduct);
+    }).toThrowErrorMatchingSnapshot();
+  });
+
+  test('it should throw an error with a tag attribute empty', () => {
+    const invalidProduct = { ...validProduct };
+    invalidProduct['tag'] = [];
+    expect(() => {
+      createProduct(invalidProduct);
+    }).toThrowErrorMatchingSnapshot();
+  });
+
+  test('it should throw an error with a price attribute negative', () => {
+    const invalidProduct = { ...validProduct };
+    invalidProduct['price'] = -1;
+    expect(() => {
+      createProduct(invalidProduct);
+    }).toThrowErrorMatchingSnapshot();
+  });
+});
+
+describe('createRandomProduct tests', () => {
+  test('it should create a product with a correct user role', () => {
+    expect(createRandomProduct('clark@kent.com')).toEqual({
```

Comment:
`toBeDefined()` instead of `toEqual()` if you are only validating the correct execution of the code